gobblin-utility/src/main/java/org/apache/gobblin/util/AvroUtils.java

```diff
@@ -935,4 +941,141 @@ public static Schema overrideNameAndNamespace(Schema input, String nameOverride,
+        }
+        else {
+          Schema copySchema = Schema.createUnion(copiedUnionTypes);
+          copyProperties(schema, copySchema);
+          return copySchema;
+        }
+      }
+      case RECORD:{
+        // check if the type of this schema matches any in the parents list
+        if (parents.stream().anyMatch(parent -> parent.fullyQualifiedType().equals(schemaEntry.fullyQualifiedType()))) {
+          fieldsWithRecursion.add(schemaEntry);
+          return null;
+        }
+        List<SchemaEntry> newParents = new ArrayList<>(parents);
+        newParents.add(schemaEntry);
+        List<Schema.Field> copiedSchemaFields = new ArrayList<>();
+        for (Schema.Field field: schema.getFields()) {
+          String fieldName = schemaEntry.fieldName != null ? schemaEntry.fieldName + "." + field.name() : field.name();
+          SchemaEntry fieldSchemaEntry = new SchemaEntry(fieldName, field.schema());
+          Schema copiedFieldSchema = dropRecursive(fieldSchemaEntry, newParents, fieldsWithRecursion);
+          if (copiedFieldSchema == null) {
+          } else {
+            Schema.Field copiedField =
+                new Schema.Field(field.name(), copiedFieldSchema, field.doc(), field.defaultValue(), field.order());
+            copyFieldProperties(field, copiedField);
+            copiedSchemaFields.add(copiedField);
+          }
+        }
+        if (copiedSchemaFields.size() > 0) {
+          Schema copiedRecord = Schema.createRecord(schema.getName(), schema.getDoc(), schema.getNamespace(),
+              schema.isError());
+          copiedRecord.setFields(copiedSchemaFields);
+          copyProperties(schema, copiedRecord);
+          return copiedRecord;
+        } else {
+          return null;
+        }
+      }
+      case ARRAY: {
+        Schema itemSchema = schema.getElementType();
+        SchemaEntry itemSchemaEntry = new SchemaEntry(schemaEntry.fieldName, itemSchema);
+        Schema copiedItemSchema = dropRecursive(itemSchemaEntry, parents, fieldsWithRecursion);
+        if (copiedItemSchema == null) {
+          return null;
+        } else {
+          Schema copiedArraySchema = Schema.createArray(copiedItemSchema);
+          copyProperties(schema, copiedArraySchema);
+          return copiedArraySchema;
+        }
+      }
+      case MAP: {
+        Schema valueSchema = schema.getValueType();
+        SchemaEntry valueSchemaEntry = new SchemaEntry(schemaEntry.fieldName, valueSchema);
+        Schema copiedValueSchema = dropRecursive(valueSchemaEntry, parents, fieldsWithRecursion);
+        if (copiedValueSchema == null) {
+          return null;
+        } else {
+          Schema copiedMapSchema = Schema.createMap(copiedValueSchema);
+          copyProperties(schema, copiedMapSchema);
+          return copiedMapSchema;
+        }
+      }
+      default: {
+        return schema;
+      }
+    }
+  }
+
+  private static void copyFieldProperties(Schema.Field field, Schema.Field copiedField) {
```

Comment:
What about other elements of `Schema` object like `doc` ?

Reply:
those are not copyable... they need to be provided when you create the field. 
You'll see that in the code above where I do: 

Schema.Field copiedField =
                new Schema.Field(field.name(), copiedFieldSchema, field.doc(), field.defaultValue(), field.order());

Will add javadoc on this method to make this clearer.